### PR TITLE
Mamba2 conversion script for original models

### LINKED ISSUE #32496: Error when converting "state-spaces/mamba2-130m" weights to huggingface-compatible format
### System Info

- Transformers version: 4.40.0

### Who can help?

@molbap @ArthurZucker

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

I tried to load https://huggingface.co/state-spaces/mamba2-130m into HF-compatible Mamba-2 (https://github.com/huggingface/transformers/pull/32080), using the [convert_mamba2_ssm_checkpoint_to_pytorch.py](https://github.com/huggingface/transformers/blob/v4.44.0/src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py) script. But the script assumes model weights to be in safetensors format:

https://github.com/huggingface/transformers/blob/984bc11b0882ff1e5b34ba717ea357e069ceced9/src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py#L32-L35

but the weight file is [is in torch bin format](https://huggingface.co/state-spaces/mamba2-130m/blob/main/pytorch_model.bin) and cannot be opened in this way.

Also, the script requires a tokenizer path:

https://github.com/huggingface/transformers/blob/984bc11b0882ff1e5b34ba717ea357e069ceced9/src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py#L55-L61

but `state-spaces/mamba2-130m` [reuses](https://github.com/state-spaces/mamba/blob/v2.2.2/benchmarks/benchmark_generation_mamba_simple.py#L37) `EleutherAI/gpt-neox-20b` tokenizer instead of having its own.

### Expected behavior

`convert_mamba2_ssm_checkpoint_to_pytorch.py` should be able to convert those Mamba-2 weights:

- https://huggingface.co/state-spaces/mamba2-130m
- https://huggingface.co/state-spaces/mamba2-370m
- https://huggingface.co/state-spaces/mamba2-780m
- https://huggingface.co/state-spaces/mamba2-1.3b
- https://huggingface.co/state-spaces/mamba2-2.7b

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Extends the Mamba2 conversion script to be compatible with the paper models and codestral. I need some help handling the tokenizer or more specifically how I can overwrite the padding side of a pretrained tokenizer and then save it with the new side.

Fixes #32496

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @molbap

### CODE REVIEW

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 1]
```diff
@@ -15,31 +15,94 @@
...
+
+    # Load and save tokenizer
+    if tokenizer_model_path is not None and not is_not_codestral:
+        tokenizer_class = LlamaTokenizerFast
+        tokenizer = tokenizer_class(tokenizer_model_path, legacy=False, from_slow=True)
+    else:
+        tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-130m-hf")
+        # TODO: how to correctly overwrite padding side so it is saved as is
+        tokenizer.padding_side = "left"
     tokenizer.save_pretrained(output_dir)
```
Comment (vasqu) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:106]: Don't know how to overwrite the padding side. Would appreciate some help here. Mamba uses the same tokenizer for all models so this should suffice.

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 2]
```diff
@@ -15,31 +15,94 @@
...
+    is_not_codestral = "dim" not in config_ssm
+
+    # Set important values from config and recalculate other resulting entries
+    hf_config.hidden_size = config_ssm["d_model"] if is_not_codestral else config_ssm["dim"]
+    hf_config.num_heads = (hf_config.hidden_size * hf_config.expand) // hf_config.head_dim
+    hf_config.num_hidden_layers = config_ssm["n_layer"] if is_not_codestral else config_ssm["n_layers"]
+    hf_config.n_groups = config_ssm.get("n_groups", 1)
+    hf_config.residual_in_fp32 = config_ssm["residual_in_fp32"]
+    hf_config.tie_word_embeddings = config_ssm["tie_embeddings"]
+    hf_config.norm_before_gate = False
```
Comment (vasqu) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:43]: This might've been a `False` default value: https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/modules/mamba2.py#L51

It's `True` as default in the config.

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 3]
```diff
@@ -15,55 +15,130 @@
...
         "-o", "--output_dir", type=str, required=True, help="Path to directory to save the converted output model to."
     )
+    parser.add_argument(
+        "-t",
+        "--tokenizer_model_path",
+        type=str,
+        default=None,
+        required=False,
+        help="Path to a tokenizer file.",
+    )
```
Comment (vasqu) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:139]: Was it pointing to https://huggingface.co/mistralai/Mamba-Codestral-7B-v0.1/blob/main/tokenizer.model.v3 ? 

Might as well automate it based on using the directory.

Comment (learning-chip) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:139]: > Was it pointing to https://huggingface.co/mistralai/Mamba-Codestral-7B-v0.1/blob/main/tokenizer.model.v3 ?

Yes, this script in master branch works out-of-box for Mamba-Codestral

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 1 above)
Comment (molbap) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:108]: Since we're using `AutoTokenizer`, I'm not sure it's possible as the parameters from this tokenizer family would be used. Llama family is left-side, but AutoTokenizer mapping will give GPTNeoX family which is right-side, so indeed saving and loading will not cut it - an alternative would be to create a `Mamba2Tokenizer` class perhaps?

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 2 above)
Comment (molbap) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:43]: Ohh nice catch! Indeed, I just tested and got a different result in the code - I believe in that case that `MambaRMSNormGated` should be updated in a different PR since only one behaviour is supported right now, RMSNorm(x) * F.silu(z) instead of RMSNorm(x * F.silu(z))

**Review by molbap — COMMENTED** [review]
Thanks a lot for working on this ! Made a couple comments and I think we need to open another PR for norm_before_gate

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:170]: Sounds good to me. Let's leave it there :+1:

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py]: I've added a `Mamba2Tokenizer` which only overrides the padding side and keeps everything else from the original `GPTNeoXTokenizer`. 

There are warnings though because I overloaded to a different class. Should I leave the warnings? And if not, is there an easy way to suppress them temporarily?

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py]: I think the current implementation follows `norm_before_gate` = `False` https://github.com/huggingface/transformers/blob/2a5a6ad18aa22e98429bb5ecb880660328030ea0/src/transformers/models/mamba2/modeling_mamba2.py#L182-L185

The original also does `RMSNorm(x * F.silu(z)`
https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/ops/triton/layernorm_gated.py#L26-L27

It will probably only be erroneous when training with the fast path on the fused kernel. But yes, a respective PR may be appropriate :D

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 4]
```diff
@@ -15,55 +15,130 @@
...
-def convert_mamba2_checkpoint_file_to_huggingface_model_file(
-    mamba2_checkpoint_path: str, tokenizer_model_path: str, output_dir: str
-) -> None:
+def convert_ssm_config_to_hf_config(config_ssm: Dict) -> Tuple[Mamba2Config, bool]:
+    """Convert a Mamba2Config from mamba_ssm to a Mamba2Config from here."""
     hf_config = Mamba2Config()
-    hf_model = Mamba2ForCausalLM(hf_config)
+
+    # Flag for codestral model
+    is_not_codestral = "dim" not in config_ssm
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:64]: Should be added now. 

Still using `is_not_codestral = "dim" not in config_ssm` to identify which model type is currently at hand. Might be changed later with a better strategy when more models come along :eyes:

**Comment by vasqu** [comment]
Tests seem unrelated to me. I'll rebase after resolving everything in the review.

**Comment by molbap** [comment]
Yes there are issues with tests on main right now, on it! unrelated to your changes. To choose the model to convert I think it should be user-dependent - meaning, it could simply be a `choice` argument in the parser and that's it, right?

**Comment by vasqu** [comment]
Oh yea for sure, I kinda just went ahead and tried to automate everything :facepalm: Lemme change it in the evening when I have proper time!

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 5]
```diff
@@ -15,55 +15,191 @@
...
+        "vocab_size": "vocab_size",
+        "pad_vocab_size_multiple": "pad_vocab_size_multiple",
+        "bos_token_id": 0,
+        "pad_token_id": 0,
+        "eos_token_id": 0,
+        "config_name": "config.json",
+        "load_state_dict": partial(load_state_dict_from_torch, ckpt_name="pytorch_model.bin"),
+        "load_and_save_tokenizer": partial(load_and_save_tokenizer, "mamba_ssm"),
+    },
+}
```
Comment (vasqu) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:125]: Using partial functions to simplify the calls down the line. With this, an extension to other model types should hopefully be easier :smile:

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -15,55 +15,191 @@
...
-        "--tokenizer_model_path",
+        "-m",
+        "--mamba2_model_type",
+        type=str,
+        default="mamba_ssm",
+        const="mamba_ssm",
+        required=True,
+        choices=("codestral", "mamba_ssm"),
+        help="The model type the conversion will be performed on. Can choose from either `codestral` or `mamba_ssm`.",
+    )
```
Comment (vasqu) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:175]: User-dependent choice on conversion. For now, defaults to `mamba_ssm` as `codestral` has a repo which can be used so I'd expect more people to use it locally for the paper models.

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py]: Has been further changed with 13889f144f1b08d407be7d7b26827aa97951324f. Supporting codestral and the original mamba ssm models.

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 5 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:125]: Possibly removing `"norm_before_gate": False,` if #32686 is merged.

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py]: Has been handled by #32686

**Comment by vasqu** [comment]
With all the fixes/patches in previous PRs, this PR should be ready now. 

I think there are still two core things to consider:
- Is the addition of the `Mamba2Tokenizer` fine?
- Is the model dictionary `_MAMBA2_MODELS_DICT` fine with the partial functions or is this overengineered? :eyes:

File: src/transformers/models/mamba2/tokenization_mamba2_fast.py
Comment (ArthurZucker) [inline src/transformers/models/mamba2/tokenization_mamba2_fast.py:1]: I don't think we need this. Overall this should still be up to the user, and the tokenizers should be set to left when converting

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 7]
```diff
@@ -15,55 +15,188 @@
...
+    # Set important values from config and recalculate other resulting entries
+    hf_config.hidden_size = config_ssm[config_dict["hidden_size"]]
+    hf_config.num_heads = (hf_config.hidden_size * hf_config.expand) // hf_config.head_dim
+    hf_config.num_hidden_layers = config_ssm[config_dict["num_hidden_layers"]]
+    hf_config.n_groups = config_ssm.get(config_dict["n_groups"], 1)
+    hf_config.residual_in_fp32 = config_ssm[config_dict["residual_in_fp32"]]
+    hf_config.tie_word_embeddings = config_ssm[config_dict["tie_word_embeddings"]]
+    hf_config.bos_token_id = config_dict["bos_token_id"]
+    hf_config.pad_token_id = config_dict["pad_token_id"]
+    hf_config.eos_token_id = config_dict["eos_token_id"]
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:60]: we could also just init with the correct values from the config_dict here. MOst of the names match so **config_dict might be simpler for things that match one to one

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 8]
```diff
@@ -15,55 +15,188 @@
...
+        "tie_word_embeddings": "tie_embeddings",
+        "vocab_size": "vocab_size",
+        "pad_vocab_size_multiple": "pad_vocab_size_multiple",
+        "bos_token_id": 0,
+        "pad_token_id": 0,
+        "eos_token_id": 0,
+        "config_name": "config.json",
+        "load_state_dict": partial(load_state_dict_from_torch, ckpt_name="pytorch_model.bin"),
+        "load_and_save_tokenizer": partial(load_and_save_tokenizer, "mamba_ssm"),
+    },
```
Comment (ArthurZucker) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:121]: for any parameters that have similar names, IMO it's useless to have them in the mapping. (like bocab size, residual etc

**Review by ArthurZucker — COMMENTED** [review]
Thanks fo the PR, let's try to aim for simplicity wherever we can!

File: src/transformers/models/mamba2/tokenization_mamba2_fast.py
Comment (vasqu) [reply src/transformers/models/mamba2/tokenization_mamba2_fast.py:1]: I originally wanted to use the gpt neox tokenizer of mamba1 and override the padding side but I wasn't aware how to do it. So when saving and reloading it is back to the non-overriden padding side. (see https://github.com/huggingface/transformers/pull/32580#discussion_r1712661263)

That's how this separate tokenizer came to be. Is there an easy way to override a padding side? So loading the tokenizer, override the padding side, and on reload the padding side should persist 👀

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 8 above)
Comment (vasqu) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:121]: True that! I will change that later

File: src/transformers/models/mamba2/tokenization_mamba2_fast.py
Comment (ArthurZucker) [reply src/transformers/models/mamba2/tokenization_mamba2_fast.py:1]: Given this: https://github.com/huggingface/transformers/blob/834ec7b1ccf9b82905d5803aca41e1744849dae7/src/transformers/tokenization_utils_base.py#L1586 I think we should update the GPTTokenizerFast to take `padding_side` as input and a test making sure loading / saving does not overwrite it!

Comment (vasqu) [reply src/transformers/models/mamba2/tokenization_mamba2_fast.py:1]: Oh wow, it's so simple :laughing: 

Removed the mamba2 tokenizer now and added `padding_side` as kwarg to the gptneox tokenizer. I tested locally to see if the padding side persists and it indeed does so. There is no test file for the gptneox tokenizer, should I create one or is there somewhere else I should write this test in?

Comment (vasqu) [reply src/transformers/models/mamba2/tokenization_mamba2_fast.py:1]: Removed the kwarg again from the tokenizer file as it's passed even without it. Testing still remains :eyes:

**Comment by ArthurZucker** [comment]
Feel free to ping me for another review once ready!

**Comment by vasqu** [comment]
@ArthurZucker Should be good for review :eyes: just check locally to make sure it works.

**Review by ArthurZucker — APPROVED** [review]
Thanks, it's a bit breaking but good for me!
